### ISSUE TITLE
Header-only library recursively including main file

### ISSUE THREAD

**Issue by RLThomaz** [clangd/clangd#337]
When building a header-only library it is very likely that we include the implementation header at the end of the main header as well as the main header at the beginning of the implementation header.

The reason is very simple, the user shall only include the main header, hence the main header is responsible for including its implementation. On the other hand, while developing the header-only library, it's necessary to include the main header file at the beginning in order to benefit from Clangd features.

However, if this approach is used, Clangd will report an error stating we cannot include the main file recursively when building a preamble.

```json
{
	"resource": "/home/user/Library/include/impl/feature_a.hpp",
	"owner": "_generated_diagnostic_collection_name_#0",
	"code": "pp_including_mainfile_in_preamble",
	"severity": 8,
	"message": "In included file: main file cannot be included recursively when building a preamble",
	"source": "clang",
	"startLineNumber": 1,
	"startColumn": 10,
	"endLineNumber": 1,
	"endColumn": 47,
	"relatedInformation": [
		{
			"startLineNumber": 328,
			"startColumn": 10,
			"endLineNumber": 328,
			"endColumn": 52,
			"message": "Error occurred here",
			"resource": "/home/user/Library/include/feature_a.hpp"
		}
	]
}
```

For example:

## Header definition

```c++
// include/feature_a.hpp
#ifndef MY_HEADER_ONLY_LIBRARY_FEATURE_A_HPP
#define MY_HEADER_ONLY_LIBRARY_FEATURE_A_HPP
...
// header definition
...
// include the implementation file
#include "impl/feature_a.hpp"
#endif
```

## Header implementation

```c++
// include/impl/feature_a.hpp
// No need for header-guards
// Include the main header definition - so we can implement using clangd features.
// Note - this is not a requirement for compilation since this is being included in the header file.
#include "feature_a.hpp"
...
// header implementation
...
```

## What I expected

Since the main header file has header-guards, clangd should check that, even though the file is being recursively included, the header-guards will stop the recursion safely.

**System information**
Output of `clangd --version`: clangd version 9.0.0-2~ubuntu18.04.2 (tags/RELEASE_900/final)
Editor/LSP plugin: VSCode with ClangD extension.
Operating system: Ubuntu 18.04

**Comment by sam-mccall** [comment]
AFAICS the error here is correct. The implementation header would be standalone if it had header guards, is there a reason you can't add them?

(Clangd doesn't support non-standalone files: see #45)

> the header-guards will stop the recursion safely.

They won't though. E.g. a class definition in the impl file will be seen twice: once via the include, and once directly.

**Comment by RLThomaz** [comment]
> AFAICS the error here is correct. The implementation header would be standalone if it had header guards, is there a reason you can't add them?

Not really, the main reason is preventing the user from including the implementation instead of the declaration header, but we could add guards for it too. I will check if adding it prevents the error, but I am not sure about it.

> They won't though. E.g. a class definition in the impl file will be seen twice: once via the include, and once directly.

But the include is guarded, which means the implementation file will only be seen once since we can only see the main header once.

**Comment by sam-mccall** [comment]
> But the include is guarded, which means the implementation file will only be seen once since we can only see the main header once.

This is true if the entry point is the public header. But if the compiler starts from the impl header, the preprocessor sees:

```
// Impl.hpp
#include "interface.h"
// Interface.hpp
#ifndef INTERFACE // taken
#define INTERFACE
// ... Contents of interface.hpp ...
#include "impl.hpp"
// Impl.hpp
#include "interface.h"
// Interface.hpp
#ifndef INTERFACE // not taken
#endif // INTERFACE
// (End of interface.hpp, back to impl)
// ... Contents of impl.hpp
// (End of impl.hpp, back to interface)
// #endif // INTERFACE
// (End of interface.hpp, back to impl)
// ... Contents of impl.hpp, again!
// (End of impl.hpp)
```

So starting from impl.hpp, the single guard on interface.hpp still allows the definitions in impl to be seen twice.

Clangd doesn't support starting from a file other than the one you opened, that's what I meant by "standalone" above.

**Comment by RLThomaz** [comment]
> This is true if the entry point is the public header. But if the compiler starts from the impl header, the preprocessor sees

Agreed, but the user is not supposed to include the implementation, so we do expect the compiler to complain.

> Clangd doesn't support starting from a file other than the one you opened, that's what I meant by "standalone" above.

Oh, now it makes a lot of sense in my head. So the issue is that we start from the file we opened, and I guess it processes it before going inside the includes, is that right? If so, it explains why I still get the error even when including header-guards on both files.

**Comment by sam-mccall** [comment]
> Oh, now it makes a lot of sense in my head. So the issue is that we start from the file we opened, and I guess it processes it before going inside the includes, is that right?

Actually no, it's supposed to process in lexical order, following includes first...

> If so, it explains why I still get the error even when including header-guards on both files.

Oh... I think this is for a different reason, an artifact of the preamble optimization. We should re-examine the behavior...

**Comment by sam-mccall** [comment]
So there are three issues here:
1. header guards are required because we only support self-contained files (#45)
2. conventional `#ifndef` guards on the main-file aren't recognized because the preamble slices them in half (#333) 
3. header guards don't suppress the error because we check in the wrong order. https://reviews.llvm.org/D78366 is out for review.

I plan to fix items 2 and 3 before closing this bug. Item 1 would be very nice to have but is a large feature requiring a lot of design, and nobody's actively working on this.

**Comment by HighCommander4** [comment]
I think this issue was closed when it shouldn't have been:

> So there are three issues here:
> 
>   [...]
> 
>    2. conventional `#ifndef` guards on the main-file aren't recognized because the preamble slices them in half ([No include insertion for symbols from the main-file of the dynamic index #333](https://github.com/clangd/clangd/issues/333))
> 
>    3. header guards don't suppress the error because we check in the wrong order. https://reviews.llvm.org/D78366 is out for review.
> 
> 
> I plan to fix items 2 and 3 before closing this bug.

The patch for item 2 is https://reviews.llvm.org/D78038, but it has not merged.

**Comment by HighCommander4** [comment]
> The patch for item 2 is https://reviews.llvm.org/D78038, but it has not merged.

I've rebased this patch and posted is as a Github PR at https://github.com/llvm/llvm-project/pull/95200

**Comment by HighCommander4** [comment]
> > The patch for item 2 is https://reviews.llvm.org/D78038, but it has not merged.
> 
> I've rebased this patch and posted is as a Github PR at [llvm/llvm-project#95200](https://github.com/llvm/llvm-project/pull/95200)

Unfortunately, local testing is showing me that the patch does not actually fix the issue discussed in this bug.

I'm not sure why that is. Maybe I made a mistake during rebasing, or something changed upstream that makes the fix approach no longer work, or there's something else I'm missing.

**Comment by matekelemen** [comment]
Is there any workaround in the meantime? Is it possible to disable preamble optimization?

**Comment by HighCommander4** [comment]
> Is there any workaround in the meantime?

Yes, please see [this comment](https://www.reddit.com/r/emacs/comments/14bf1uw/comment/lj5ojt2/).
